sky/jobs/server/core.py

```diff
@@ -367,107 +366,155 @@ def launch(
                 'job_controller_indicator_file':
                     managed_job_constants.JOB_CONTROLLER_INDICATOR_FILE,
+                'pre_created_job_ids': pre_created_job_ids,
+                'num_jobs': num_jobs,
                 **controller_utils.shared_controller_vars_to_fill(
                     controller,
                     remote_user_config_path=remote_user_config_path,
-                    # TODO(aylei): the mutated config will not be updated
-                    # afterwards without recreate the controller. Need to
-                    # revisit this.
                     local_user_config=mutated_user_config,
                 ),
             }
 
-            yaml_path = os.path.join(
-                managed_job_constants.JOBS_CONTROLLER_YAML_PREFIX,
-                f'{name}-{dag_uuid}-{consolidation_mode_job_id}-{job_rank}.yaml'
-            )
+            if pre_created_job_ids is not None:
+                vars_to_fill['pre_created_job_ids'] = pre_created_job_ids
+                # Create job_id_to_rank dictionary by sorting job IDs and
+                # assigning ranks. The last job ID (controller task's ray
+                # job ID) will be added in the template
+                sorted_job_ids = sorted(pre_created_job_ids)
+                job_id_to_rank = {
+                    str(job_id): rank
+                    for rank, job_id in enumerate(sorted_job_ids)
+                }
+                vars_to_fill['job_id_to_rank'] = job_id_to_rank
+                yaml_path = os.path.join(
+                    managed_job_constants.JOBS_CONTROLLER_YAML_PREFIX,
+                    f'{name}-{dag_uuid}-multi.yaml')
+                logger.info(
+                    f'{colorama.Fore.YELLOW}'
+                    f'Launching {num_jobs} managed jobs {dag.name!r} '
+                    f'with IDs {pre_created_job_ids} from jobs controller...'
+                    f'{colorama.Style.RESET_ALL}')
+            else:
+                if consolidation_mode_job_ids is not None:
+                    sorted_job_ids = sorted(consolidation_mode_job_ids)
+                    job_id_to_rank = {
+                        str(job_id): rank
+                        for rank, job_id in enumerate(sorted_job_ids)
+                    }
+                    vars_to_fill['job_id_to_rank'] = job_id_to_rank
+                yaml_path = os.path.join(
+                    managed_job_constants.JOBS_CONTROLLER_YAML_PREFIX,
+                    f'{name}-{dag_uuid}.yaml')
+
+                job_identity = ''
+                if job_rank is not None:
+                    job_identity = f' (rank: {job_rank})'
+                job_controller_postfix = (' from jobs controller'
+                                          if consolidation_mode_job_ids is None
+                                          else '')
+                logger.info(
+                    f'{colorama.Fore.YELLOW}'
+                    f'Launching managed job {dag.name!r}{job_identity}'
```

Comment:
Should technically also be plural if consolidation_mode_job_ids has length >0
I get the feeling this logic can be simplified a lot.